### ISSUE TITLE
kube-apiserver v1.19 fails to exec the preStop lifecycle hook

### ISSUE THREAD

**Issue by ialidzhikov** [gardener/gardener#3278]
**How to categorize this issue?**

/kind bug
/priority normal

**What happened**:
https://github.com/gardener/gardener/pull/1275/files adds a prestop hook that relies on presence of `sh` in the apiserver image.
`sh` seems to be missing in `k8s.gcr.io/kube-apiserver:v1.19.4`:
```
$ docker run -ti k8s.gcr.io/kube-apiserver:v1.19.4 sh
2020/12/08 15:34:45 Running command:
Command env: (log-file=, also-stdout=false, redirect-stderr=true)
Run from directory:
Executable path: sh
Args (comma-delimited): sh
2020/12/08 15:34:45 starting command: exec: "sh": executable file not found in $PATH
```

And hence on each rollout we have events that report failure of the prestop hook:

```
53m         Warning   FailedPreStopHook                 pod/kube-apiserver-59d8cbdf5b-9cfkd                          Exec lifecycle hook ([sh -c sleep 5]) for Container "kube-apiserver" in Pod "kube-apiserver-59d8cbdf5b-9cfkd_shoot--foo--bar(01fa2ddb-8156-4e62-aba0-716343bba136)" failed - error: command 'sh -c sleep 5' exited with 126: , message: "OCI runtime exec failed: exec failed: container_linux.go:370: starting container process caused: exec: \"sh\": executable file not found in $PATH: unknown\r\n"
```

`sh` seems to be present on v1.18 and below.

**What you expected to happen**:
The apiserver preStop hook to do not fail for v1.19.

- Gardener version: v1.13.2
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

**Comment by vpnachev** [comment]
Good catch, the only binary inside the image, other than kube-apiserver, is `/usr/sbin/tzconfig`.
This should be related to the migration to distroless https://github.com/kubernetes/kubernetes/pull/90674 in 1.19.

**Comment by timuthy** [comment]
The prestop hook should not be needed since we can use `/readyz` with `shutdown-delay-duration` flag (/cc @amshuman-kr)

```
--shutdown-delay-duration duration
Time to delay the termination. During that time the server keeps serving requests normally. The endpoints /healthz and /livez will return success, but /readyz immediately returns failure. Graceful termination starts after this delay has elapsed. This can be used to allow load balancer to stop sending traffic to this server.
```

**Comment by amshuman-kr** [comment]
> The prestop hook should not be needed since we can use `/readyz` with `shutdown-delay-duration` flag (/cc @amshuman-kr)
> 
> ```
> --shutdown-delay-duration duration
> Time to delay the termination. During that time the server keeps serving requests normally. The endpoints /healthz and /livez will return success, but /readyz immediately returns failure. Graceful termination starts after this delay has elapsed. This can be used to allow load balancer to stop sending traffic to this server.
> ```

That should be sufficient, yes 👍

**Comment by amshuman-kr** [comment]
But I hope the following issue is not a problem because we have configured the delay to `20s`.
https://github.com/kubernetes/kubernetes/issues/88293

**Comment by timuthy** [comment]
Good point @amshuman-kr. @ialidzhikov are you already working on a fix or do you want someone else to take over?

**Comment by ialidzhikov** [comment]
> The prestop hook should not be needed since we can use `/readyz` with `shutdown-delay-duration` flag (/cc @amshuman-kr)
> 
> ```
> --shutdown-delay-duration duration
> Time to delay the termination. During that time the server keeps serving requests normally. The endpoints /healthz and /livez will return success, but /readyz immediately returns failure. Graceful termination starts after this delay has elapsed. This can be used to allow load balancer to stop sending traffic to this server.
> ```

But `/readyz` and `--shutdown-delay-duration ` are only available for kubernetes apiserver >= v1.16. So do you mean to use `--shutdown-delay-duration ` for k8s >= v1.16 and keep the preStop hook for < v1.16?
I am not looking into this currently. Feel free to take it if you have something in mind.

 I generally don't understand what is the rational behind this preStop hook. @amshuman-kr , what is this?

> The delay in preStop help give the kube-proxies a chance to
propagate the changes to the iptables in all the nodes as noted [here](https://github.com/kubernetes/kubernetes/issues/43576).

**Comment by timuthy** [comment]
> But /readyz and --shutdown-delay-duration are only available for kubernetes apiserver >= v1.16. So do you mean to use --shutdown-delay-duration for k8s >= v1.16 and keep the preStop hook for < v1.16?

Yes, that's the idea.

> I generally don't understand what is the rational behind this preStop hook. @amshuman-kr , what is this?

The `preStop` is meant to let the Kube-Apiserver accept incoming requests for a few seconds (after termination was triggered) because otherwise the Pod might be gone while the traffic is still routed to the Pod (delay in service/endpoint update).

**Comment by amshuman-kr** [comment]
> I generally don't understand what is the rational behind this preStop hook. @amshuman-kr , what is this?

Yes, @timuthy is correct. The idea is to give time for the traffic routing to take effect. This mitigates losing requests during rolling updates https://github.com/kubernetes/kubernetes/issues/43576.
